### PR TITLE
feat: a Rust e2e test using TestCluster

### PR DESCRIPTION
**Pull request by gdanezis**
Added a rust end-to-end test between Rust and Move contracts:
* Compile and upload the contract.
* Then check events are fired

### CODE REVIEW

File: crates/walrus-e2e-tests/Cargo.toml
Comment (mlegner) [inline crates/walrus-e2e-tests/Cargo.toml:1]: Suggestion: Add `default-members = ["crates/walrus-core", "crates/walrus-service", "crates/walrus-test-utils"]` to the workspace `Cargo.toml`. Then, standard Cargo commands like `cargo build` or `cargo test` only act on those by default. Testing the E2E tests is then possible with `cargo test --package walrus-e2e-tests`.

File: contracts/blob_store/sources/storage_accounting.move [hunk 1]
```diff
@@ -11,14 +11,14 @@ module blob_store::storage_accounting {
 
     /// Holds information about a future epoch, namely how much
     /// storage needs to be reclaimed and the rewards to be distributed.
-    struct FutureAccounting<phantom TAG, phantom WAL:store> has store {
+    struct FutureAccounting<phantom TAG, phantom WAL> has store {
```
Comment (mlegner) [inline contracts/blob_store/sources/storage_accounting.move]: Question: Why are those changes needed?

File: crates/walrus-e2e-tests/src/lib.rs
Comment (mlegner) [inline crates/walrus-e2e-tests/src/lib.rs:1]: Suggestion: Move all E2E tests to their separate file in `crates/walrus-e2e-tests/tests`, which is the general convention (see [here](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests)). The crate itself can then simply contain some utility functions shared by all the tests (like the `compile_package` function).

File: crates/walrus-e2e-tests/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+        let mut test_cluster = TestClusterBuilder::new().build().await;
+        let context = &mut test_cluster.wallet;
+        let sender = context.active_address().unwrap();
+
+        let mut current_dir = std::env::current_dir()?;
+        // navigate to ../../contracts/blob_store
+        current_dir.pop();
+        current_dir.pop();
+        current_dir.push("contracts");
+        current_dir.push("blob_store");
```
Comment (mlegner) [inline crates/walrus-e2e-tests/src/lib.rs]: Nit: The following feels more natural to me:
```suggestion
        // navigate to ../../contracts/blob_store
        let package_dir = std::env::current_dir()?
            .parent()
            .unwrap()
            .parent()
            .unwrap()
            .join("contracts")
            .join("blob_store");
```

File: crates/walrus-e2e-tests/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        let events = sui
+            .event_api()
+            .query_events(EventFilter::Sender(sender), None, None, false)
+            .await?;
+
+        assert!(
+            events
+                .data
+                .iter()
+                .any(|e| e.type_.name.clone().into_string() == "EpochChangeSync"),
```
Comment (mlegner) [inline crates/walrus-e2e-tests/src/lib.rs]: Nit:
```suggestion
                .any(|e| e.type_.name.to_string() == "EpochChangeSync"),
```

File: contracts/blob_store/sources/system.move [hunk 4]
```diff
@@ -135,15 +136,26 @@ module blob_store::system {
         }
     }
 
+    public fun share_new<TAG, WAL>(
```
Comment (mlegner) [inline contracts/blob_store/sources/system.move:141]: Remark: There is a linter warning about this. I'm not sure if there is a better way; if not, I suggest to add an exception here:
```suggestion
    #[allow(lint(share_owned))]
    public fun share_new<TAG, WAL>(
```

File: crates/walrus-e2e-tests/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        assert!(
+            events
+                .data
+                .iter()
+                .any(|e| e.type_.name.clone().into_string() == "EpochChangeDone"),
+            "No EpochChangeDone event found"
+        );
+
+        // Print events
+        println!("Events: {:?}", events);
```
Comment (mlegner) [inline crates/walrus-e2e-tests/src/lib.rs]: Suggestion: As test output is normally only shown on failure, I'd suggest moving this above the asserts to actually see it in those cases.

File: crates/walrus-e2e-tests/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            "No EpochChangeSync event found"
+        );
+
+        assert!(
+            events
+                .data
+                .iter()
+                .any(|e| e.type_.name.clone().into_string() == "EpochChangeDone"),
+            "No EpochChangeDone event found"
+        );
```
Comment (mlegner) [inline crates/walrus-e2e-tests/src/lib.rs]: Nit: How about this version that can be extended easily to more event types?
```suggestion
        ["EpochChangeSync", "EpochChangeDone"]
            .into_iter()
            .for_each(|event_type| {
                assert!(
                    events
                        .data
                        .iter()
                        .any(|e| e.type_.name.to_string() == event_type),
                    "No {} event found",
                    event_type
                )
            });
```

File: contracts/blob_store/sources/storage_accounting.move (hunk 1 above)
Comment (gdanezis) [reply contracts/blob_store/sources/storage_accounting.move]: These were artefacts of not merging the changes from the previous commit to main. Now they are gone, since they were applied before.

File: crates/walrus-e2e-tests/Cargo.toml
Comment (gdanezis) [reply crates/walrus-e2e-tests/Cargo.toml:1]: Excellent idea!

File: crates/walrus-e2e-tests/src/lib.rs
Comment (gdanezis) [reply crates/walrus-e2e-tests/src/lib.rs:1]: Done!

File: crates/walrus-e2e-tests/src/lib.rs (hunk 5 above)
Comment (gdanezis) [reply crates/walrus-e2e-tests/src/lib.rs]: I removed all prints, there were for debugging.

File: contracts/blob_store/sources/system.move (hunk 4 above)
Comment (gdanezis) [reply contracts/blob_store/sources/system.move:141]: Ok, checked this is ok. Added a lint.

File: crates/walrus-e2e-tests/src/lib.rs (hunk 6 above)
Comment (gdanezis) [reply crates/walrus-e2e-tests/src/lib.rs]: Actually I am going to do a find, to extract the package ID here too.

File: crates/walrus-e2e-tests/src/lib.rs
Comment (mlegner) [reply crates/walrus-e2e-tests/src/lib.rs:1]: I've pushed a commit with some slight reorganization in c456b8211a46009df635cf979f301ef9268e0920.

If you agree, please go ahead and merge. LGTM from my side. 🙂

Comment (gdanezis) [reply crates/walrus-e2e-tests/src/lib.rs:1]: Thats great!
